migrator/zenodo_rdm_migrator/extract/kafka.py

```diff
@@ -7,33 +7,291 @@
+
+    def iter_tx_info(self):
+        """Yield commited transactions info."""
+        consumer = self._tx_consumer
+        for tx_msg in consumer:
+            self._dump_msg(self.tx_topic, tx_msg)
+
+            if not tx_msg.value:
+                # Sometimes messages don't contain a value... So far it's not been an
+                # issue, but maybe logging in DEBUG could help at some point.
+                self.logger.debug(f"No message value for tx_info {tx_msg}")
+                consumer.commit()
+                continue
+
+            tx_id, lsn = map(int, tx_msg.value["id"].split(":"))
+            # We don't yield anything before the configured last transaction ID
+            if tx_id <= self.last_tx:
+                consumer.commit()
+                continue
+            if tx_msg.value["status"] == "BEGIN":
+                # ignore BEGIN statements
+                consumer.commit()
+                continue
+            elif tx_msg.value["status"] == "END":
+                consumer.commit()
+                yield (tx_id, tx_msg.value)
+
+    def iter_ops(self):
+        """Yields operations/statements."""
+        consumer = self._ops_consumer
+        for op_msg in consumer:
+            self._dump_msg(self.ops_topic, op_msg)
+
+            if not op_msg.value:
+                self.logger.debug(f"No message value for op {op_msg}")
+                consumer.commit()
+                continue
+            tx_id = op_msg.value["source"]["txId"]
+            if tx_id <= self.last_tx:
+                consumer.commit()
+                continue
+            consumer.commit()
+            yield (tx_id, dict(key=op_msg.key, **op_msg.value))
+
+    def _yield_completed_tx(self, min_batch=None):
+        """Yields completed transactions."""
+        completed_tx_batch = []
+        for tx_state in self.tx_registry.values():
+            if not tx_state.complete:
+                # We stop at the first non-completed transaction
+                break
+            completed_tx_batch.append(tx_state)
+
+            # if we reached the minimum batch size, we stop
+            if min_batch and len(completed_tx_batch) == min_batch:
+                break
+
+        # If we didn't make a big enough batch we return
+        if min_batch and len(completed_tx_batch) < min_batch:
+            return
+
+        for tx in completed_tx_batch:
+            del self.tx_registry[tx.id]
+            # Keep track of the last yielded transaction ID
+            self._last_yielded_tx = tx.id
+            yield Tx(id=tx.id, operations=list(tx.ops))
 
     def run(self):
-        """Yield one element at a time."""
-        consumer = KafkaConsumer(*self.topics, **self.kafka_cfg)
-        yield from consumer
+        """Return a blocking generator yielding completed transactions."""
+        # We're using an (always) SortedDict, since we want to yield transactions in
+        # their correct order.
+        try:
```

Comment:
understanding check:
1. we consume all available info (does this ever stop?)
2. we consume all the available ops
3. we yield those that completed
4. back to step 1

I wonder if there is risk of starvation in any of the steps (i.e. not passing to the subsequent).

Reply:
> 1. we consume all available info (does this ever stop?)

Correct. Once we go through the "bulk" of it (depends on the starting point/offset/timestamp), it starts behaving more like a live stream that returns continuous batches (it stops iterating if there are no messages after 3sec, as per the `consumer_timeout_ms` config). 

I'm actually considering adding a "max messages" param to chunk the delta runs, because the first time we run a delta (i.e. after the snapshot has been loaded) it will be ~8h worth of messages, which will be stored in memory... So pretty good question/catch :)

> 2. we consume all the available ops
> 3. we yield those that completed

important: we yield the "earliest" transactions that have been completed. I.e. imagine:

1. We got Tx1, Tx2, and Tx3
2. Tx2 and Tx3 are complete, but Tx1 still has ops (that will come in a later iteration of the consumer)
3. In this case we don't yield Tx2 and Tx3, since we're still waiting for Tx1 to complete (hopefully in the next iteration).

> 4. back to step 1

Correct!

> I wonder if there is risk of starvation in any of the steps (i.e. not passing to the subsequent).

See the point regarding the `consumer_timeout_ms` above ☝️

Reply:
This is very helpful, thanks for taking the time to write details comments. I wonder if we should put this in the class docstring? since it is not trivial to understand it might help others.

Reply:
Added the max fetching parameters, and some additional docstring comments in https://github.com/zenodo/zenodo-rdm/pull/454/commits/d93da36210d4211efeb7146752a965bf60eda62a